libfwupdplugin/fu-mei-device.c

```diff
@@ -0,0 +1,315 @@
+ * @self: a #FuMeiDevice
+ *
+ * Gets the maximum message length.
+ *
+ * Returns: integer
+ *
+ * Since: 1.8.2
+ **/
+guint32
+fu_mei_device_get_max_msg_length(FuMeiDevice *self)
+{
+	FuMeiDevicePrivate *priv = GET_PRIVATE(self);
+	g_return_val_if_fail(FU_IS_MEI_DEVICE(self), G_MAXUINT32);
+	return priv->max_msg_length;
+}
+
+/**
+ * fu_mei_device_get_protocol_version:
+ * @self: a #FuMeiDevice
+ *
+ * Gets the protocol version, or 0x for unset.
+ *
+ * Returns: integer
+ *
+ * Since: 1.8.2
+ **/
+guint8
+fu_mei_device_get_protocol_version(FuMeiDevice *self)
+{
+	FuMeiDevicePrivate *priv = GET_PRIVATE(self);
+	g_return_val_if_fail(FU_IS_MEI_DEVICE(self), G_MAXUINT8);
+	return priv->protocol_version;
+}
+
+/**
+ * fu_mei_device_connect:
+ * @self: a #FuMeiDevice
+ * @guid: A GUID, e.g. "2800f812-b7b4-2d4b-aca8-46e0ff65814c"
+ * @req_protocol_version: required protocol version, or 0
+ * @error: (nullable): optional return location for an error
+ *
+ * Connects to the MEI device for a specific GUID.
+ *
+ * Returns: %TRUE for success
+ *
+ * Since: 1.8.2
+ **/
+gboolean
+fu_mei_device_connect(FuMeiDevice *self,
+		      const gchar *guid,
+		      guchar req_protocol_version,
+		      GError **error)
+{
+#ifdef HAVE_MEI_H
+	FuMeiDevicePrivate *priv = GET_PRIVATE(self);
+	fwupd_guid_t guid_le = {0x0};
+	struct mei_client *cl;
+	struct mei_connect_client_data data = {0x0};
+
+	g_return_val_if_fail(FU_IS_MEI_DEVICE(self), FALSE);
+	g_return_val_if_fail(guid != NULL, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	if (!fwupd_guid_from_string(guid, &guid_le, FWUPD_GUID_FLAG_NONE, error))
+		return FALSE;
+	memcpy(&data.in_client_uuid, &guid_le, sizeof(guid_le));
+	if (!fu_udev_device_ioctl(FU_UDEV_DEVICE(self),
+				  IOCTL_MEI_CONNECT_CLIENT,
+				  (guint8 *)&data,
+				  NULL, /* rc */
+				  FU_MEI_DEVICE_IOCTL_TIMEOUT,
+				  error))
+		return FALSE;
+
+	cl = &data.out_client_properties;
+	if (req_protocol_version > 0 && cl->protocol_version != req_protocol_version) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "Intel MEI protocol version not supported %i",
+			    cl->protocol_version);
+		return FALSE;
+	}
+
+	/* success */
+	priv->max_msg_length = cl->max_msg_length;
+	priv->protocol_version = cl->protocol_version;
+	return TRUE;
+#else
+	g_set_error_literal(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "linux/mei.h not supported");
+	return FALSE;
+#endif
+}
+
+/**
+ * fu_mei_device_read:
+ * @self: a #FuMeiDevice
+ * @buf: (out): data
+ * @bufsz: size of @data
+ * @bytes_read: (nullable): bytes read
+ * @timeout_ms: timeout
+ * @error: (nullable): optional return location for an error
+ *
+ * Read raw bytes from the device.
+ *
+ * Returns: %TRUE for success
+ *
+ * Since: 1.8.2
+ **/
+gboolean
+fu_mei_device_read(FuMeiDevice *self,
+		   guint8 *buf,
+		   gsize bufsz,
+		   gsize *bytes_read,
+		   guint timeout_ms,
+		   GError **error)
```

Comment:
If this is coming into plugin ABI, do we need some sort of mutex to ensure two plugins haven't stepped on each other's data in multiple back to back reads?

Reply:
The way this will work in practice is that the AMT device is a superclass of FuMeiDevice -- and the possible plugins will create that specific instance, we run probe, open, setup, close, then move onto the next device. Each instance is namespaced by the connection GUID which seems to provide some kind of interface API (hence the GUID on the ioctl _connect_). Of course, two things can't use the same /dev/mei0 interface, but I don't think it's allowed to happen. If we started threading the plugin coldplug (which I don't think we should do) then we'd need some kind of access control, but in this case I'm tempted to KISS and then if if we do need two things writing to the same device at the same time create a `fu-mei-backend.c` which handles the device open/closes and does the locking there.

Reply:
Yeah I'm aware we're not planning to thread coldplug, just being cognizant that if we want to some day that we need to watch out for design problems.  Sounds like a good approach.

Reply:
I'll merge this, and then propose the new hardware support next week so we can see if there are any changes needed in the baseclass before the next tarball. Thanks for all the review.